File: backend-java/src/main/java/com/workflow/service/LlmTestService.java
package com.workflow.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.workflow.config.LlmProviderConfig;
import com.workflow.util.ErrorMessages;
import com.workflow.util.LlmConfigUtils;
import com.workflow.util.ObjectUtils;
import com.workflow.util.LlmErrorResponseBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.List;
import java.util.Map;

/**
 * Service for testing LLM provider connections.
 * Uses provider registry (OCP), shared HTTP helper (DRY), and configurable URLs.
 */
@Service
public class LlmTestService {
    private static final Logger log = LoggerFactory.getLogger(LlmTestService.class);

    private final LlmProviderConfig.LlmProviderUrls providerUrls;
    private final Map<String, ProviderTester> providerRegistry;
    private final ObjectMapper objectMapper;

    @FunctionalInterface
    private interface ProviderTester {
        Map<String, String> test(String apiKey, String baseUrl, String model);
    }

    public LlmTestService(LlmProviderConfig.LlmProviderUrls providerUrls, ObjectMapper objectMapper) {
        this.providerUrls = providerUrls;
        this.objectMapper = objectMapper;
        providerRegistry = Map.of(
                "openai", this::testOpenAi,
                "anthropic", this::testAnthropic,
                "gemini", this::testGemini,
                "custom", this::testCustom
        );
    }

    /**
     * Test provider by type - OCP: new providers via registry.
     */
    public Map<String, String> testProvider(String type, String apiKey, String baseUrl, String model) {
        var tester = providerRegistry.get(ObjectUtils.orDefaultIfBlank(type, "").toLowerCase());
        if (tester == null) {
            return LlmErrorResponseBuilder.error(ErrorMessages.unknownProviderType(type));
        }
        return tester.test(apiKey, baseUrl, model);
    }

    public Map<String, String> testOpenAi(String apiKey, String baseUrl, String model) {
        String url = LlmConfigUtils.buildChatCompletionsUrl(LlmConfigUtils.normalizeBaseUrl(baseUrl, providerUrls.getOpenai()));
        return testOpenAiCompatible(url, apiKey, model);
    }

    public Map<String, String> testAnthropic(String apiKey, String baseUrl, String model) {
        String url = LlmConfigUtils.normalizeBaseUrl(baseUrl, providerUrls.getAnthropic()) + "/messages";
        return httpPost(url, Map.of(
                "x-api-key", apiKey,
                "anthropic-version", "2023-06-01",
                "Content-Type", "application/json"
        ), buildOpenAiCompatiblePayload(model, 5));
    }

    public Map<String, String> testGemini(String apiKey, String baseUrl, String model) {
        String bUrl = LlmConfigUtils.normalizeBaseUrl(baseUrl, providerUrls.getGemini());
        String encodedModel = URLEncoder.encode(model != null ? model : "", StandardCharsets.UTF_8);
        String encodedKey = URLEncoder.encode(apiKey != null ? apiKey : "", StandardCharsets.UTF_8);
        String url = bUrl + "/models/" + encodedModel + ":generateContent?key=" + encodedKey;
        return httpPost(url, Map.of("Content-Type", "application/json"),
                buildGeminiPayload(5));
    }

    public Map<String, String> testCustom(String apiKey, String baseUrl, String model) {
        if (baseUrl == null || baseUrl.isBlank()) {
            return LlmErrorResponseBuilder.error(ErrorMessages.BASE_URL_REQUIRED_CUSTOM);
        }
        String url = LlmConfigUtils.buildChatCompletionsUrl(baseUrl);
        return testOpenAiCompatible(url, apiKey, model);
    }

    private Map<String, String> testOpenAiCompatible(String url, String apiKey, String model) {
        return httpPost(url, Map.of(
                "Authorization", "Bearer " + apiKey,
                "Content-Type", "application/json"
        ), buildOpenAiCompatiblePayload(model, 5));
    }

    private String buildOpenAiCompatiblePayload(String model, int maxTokens) {
        try {
            Map<String, Object> payload = Map.of(
                    "model", ObjectUtils.orDefaultIfBlank(model, ""),
                    "messages", List.of(Map.of("role", "user", "content", "Hello")),
                    "max_tokens", maxTokens
            );
            return objectMapper.writeValueAsString(payload);
        } catch (Exception e) {
            log.warn("Failed to build OpenAI payload: {}", e.getMessage());
            throw new IllegalStateException(ErrorMessages.UNEXPECTED_ERROR);
        }
    }

    private String buildGeminiPayload(int maxOutputTokens) {
        try {
            Map<String, Object> payload = Map.of(
                    "contents",
                    List.of(
                            Map.of(
                                    "role",
                                    "user",
                                    "parts",
                                    List.of(Map.of("text", "Hello")))),
                    "generationConfig", Map.of("maxOutputTokens", maxOutputTokens));
            return objectMapper.writeValueAsString(payload);
        } catch (Exception e) {
            log.warn("Failed to build Gemini payload: {}", e.getMessage());
            throw new IllegalStateException(ErrorMessages.UNEXPECTED_ERROR);
        }
    }

    /**
     * Shared HTTP POST helper (DRY).
     */
    private Map<String, String> httpPost(String url, Map<String, String> headers, String body) {
        try {
            var builder = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .POST(HttpRequest.BodyPublishers.ofString(body));
            for (var e : headers.entrySet()) {
                builder.header(e.getKey(), e.getValue());
            }
            var request = builder.build();
            var response = HttpClient.newHttpClient().send(request, HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() == 200) {
                return Map.of("status", "success", "message", ErrorMessages.CONNECTED_SUCCESSFULLY);
            }
            log.warn("LLM test API error {}: {}", response.statusCode(), response.body());
            return LlmErrorResponseBuilder.error(ErrorMessages.llmApiError(response.statusCode()));
        } catch (Exception e) {
            log.warn("LLM test HTTP request failed: {}", e.getMessage());
            return LlmErrorResponseBuilder.error(ErrorMessages.UNEXPECTED_ERROR);
        }
    }
}

File: backend/services/llm_test_service.py
"""
LLM Test Service - DRY shared logic for testing LLM provider connections.
Uses shared HTTP helper for all providers (DRY).
"""
from typing import Dict, Any, Optional
import httpx

from ..utils.logger import get_logger

logger = get_logger(__name__)

_DEFAULT_STATUS_MESSAGES: Dict[int, str] = {
    400: "API error",
    401: "Invalid API key (401 Unauthorized)",
    404: "Not found (404)",
    429: "Rate limit exceeded (429)",
}


async def _do_http_post(
    url: str,
    headers: Dict[str, str],
    body: Dict[str, Any],
    url_display: Optional[str] = None,
    status_overrides: Optional[Dict[int, str]] = None,
) -> Dict[str, str]:
    """
    Shared HTTP POST helper for provider tests (DRY).
    Handles connect, timeout, and response status. status_overrides can customize error messages.
    """
    display = url_display or url
    overrides = status_overrides or {}
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(url, headers=headers, json=body)
            if response.status_code == 200:
                return {"status": "success", "message": "Connected successfully"}
            msg = overrides.get(response.status_code)
            if msg is None:
                if response.status_code == 400:
                    try:
                        error_data = response.json() if response.text else {}
                        error_msg = error_data.get("error", {}).get("message", response.text[:200])
                        msg = f"API error: {error_msg}"
                    except Exception:
                        msg = "API error"
                else:
                    msg = _DEFAULT_STATUS_MESSAGES.get(
                        response.status_code,
                        f"API returned status {response.status_code}: {response.text[:200]}",
                    )
            return {"status": "error", "message": msg}
    except httpx.ConnectError as e:
        return {"status": "error", "message": f"Cannot connect to {display}: {str(e)}"}
    except httpx.TimeoutException:
        return {"status": "error", "message": "Connection timed out after 30 seconds"}
    except Exception as e:
        return {"status": "error", "message": f"Error: {str(e)}"}


async def test_openai(base_url: str, api_key: str, model: str) -> Dict[str, str]:
    """Test OpenAI API connection."""
    url = (base_url or "https://api.openai.com/v1") + "/chat/completions"
    return await _test_openai_compatible(url, api_key, model)


async def test_anthropic(base_url: str, api_key: str, model: str) -> Dict[str, str]:
    """Test Anthropic API connection."""
    url = (base_url or "https://api.anthropic.com/v1") + "/messages"
    return await _do_http_post(
        url,
        headers={
            "x-api-key": api_key,
            "anthropic-version": "2023-06-01",
            "Content-Type": "application/json",
        },
        body={
            "model": model,
            "messages": [{"role": "user", "content": "Hello"}],
            "max_tokens": 5,
        },
    )


async def test_gemini(base_url: str, api_key: str, model: str) -> Dict[str, str]:
    """Test Google Gemini API connection (AI Studio key) or Vertex :generateContent with ADC when key is empty."""
    body = {
        "contents": [{"parts": [{"text": "Hello"}]}],
        "generationConfig": {"maxOutputTokens": 5},
    }
    if (api_key or "").strip():
        url = (base_url or "https://generativelanguage.googleapis.com/v1beta") + f"/models/{model}:generateContent?key={api_key}"
        return await _do_http_post(
            url,
            headers={"Content-Type": "application/json"},
            body=body,
            url_display=base_url or "Gemini API",
        )
    try:
        from ..utils.vertex_gemini import post_vertex_generate_content, vertex_ai_configured
    except Exception as e:
        return {"status": "error", "message": f"Vertex AI not available: {e}"}
    if not vertex_ai_configured():
        return {
            "status": "error",
            "message": "No Gemini API key and Vertex AI is not configured (set GOOGLE_CLOUD_PROJECT or GCP_PROJECT).",
        }
    try:
        resp = await post_vertex_generate_content(model, body, timeout=30.0)
    except Exception as e:
        return {"status": "error", "message": f"Vertex AI (ADC) error: {e}"}
    if resp.status_code == 200:
        return {"status": "success", "message": "Connected successfully (Vertex AI with ADC)"}
    return {
        "status": "error",
        "message": f"Vertex AI returned status {resp.status_code}: {(resp.text or '')[:200]}",
    }


async def test_custom(base_url: str, api_key: str, model: str) -> Dict[str, str]:
    """Test custom OpenAI-compatible API connection."""
    if not base_url:
        from fastapi import HTTPException
        raise HTTPException(status_code=400, detail="base_url is required for custom providers")
    url = base_url.rstrip("/") + "/chat/completions"
    return await _test_openai_compatible(url, api_key, model)


async def _test_openai_compatible(url: str, api_key: str, model: str) -> Dict[str, str]:
    """Shared logic for OpenAI and custom (OpenAI-compatible) providers."""
    result = await _do_http_post(
        url,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        body={
            "model": model,
            "messages": [{"role": "user", "content": "Hello"}],
            "max_tokens": 5,
        },
        status_overrides={404: f"Model '{model}' not found (404)"},
    )
    if result["status"] == "success":
        result["message"] = "Connected successfully! ✓"
    return result
